bsm_helper/src/bsm_helper.cpp

```diff
@@ -1,23 +1,30 @@
+/* Copyright (C) 2021 LEIDOS.
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ use this file except in compliance with the License. You may obtain a copy of
+ the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ License for the specific language governing permissions and limitations under
+ the License.
+*/
 #include "bsm_helper/bsm_helper.h"
 
 namespace BSMHelper{
 
 
     std::string BSMHelper::bsmIDtoString(std::vector<uint8_t, std::allocator<uint8_t>> id)
     {
-
-        std::string id_test;
-
-        for(size_t i =0; i < id.size();i++)
+        if(id.size() > 4)
         {
-            id_test.append(std::to_string(id.at(i)));
+            throw std::invalid_argument("Invalid id value");
         }
 
-        std::stringstream int_id(id_test);
-
-        unsigned long id_num;
-
-        int_id >> id_num;
+        unsigned long id_num = 0;
+        for(size_t i = 0; i < id.size(); i++)
+        {
+          id_num = (id_num * pow(10, 2)) + id.at(i) ;
```

Comment:
I think this method is prone to unit conversion errors. The pow function returns a double which means id_num might be upcast first which could result in a bit flip. At least that's what my intuition tells me. I thick bitwise operators are the better choice here

```
for(...)
{
   id_num = (id_num << 8) | id.at(i);
}
```

See
https://www.geeksforgeeks.org/bitwise-operators-in-c-cpp/